pkg/phlaredb/profile_store.go

```diff
@@ -0,0 +1,407 @@
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// create index
+	s.index, err = newProfileIndex(32, s.metrics)
+	if err != nil {
+		return err
+	}
+
+	s.path = path
+	s.cfg = cfg
+
+	s.slice = s.slice[:0]
+
+	s.rowsFlushed = 0
+
+	return nil
+}
+
+func (s *profileStore) Close() error {
+	return nil
+}
+
+func (s *profileStore) RowGroups() (rowGroups []parquet.RowGroup) {
+
+	rowGroups = make([]parquet.RowGroup, len(s.rowGroups))
+	for pos := range rowGroups {
+		rowGroups[pos] = s.rowGroups[pos]
+	}
+	return rowGroups
+}
+
+func (s *profileStore) profileSort(i, j int) bool {
+	// first compare the labels, if they don't match return
+	var (
+		pI   = s.slice[i]
+		pJ   = s.slice[j]
+		lbsI = s.index.profilesPerFP[pI.SeriesFingerprint].lbs
+		lbsJ = s.index.profilesPerFP[pJ.SeriesFingerprint].lbs
+	)
+	if cmp := phlaremodel.CompareLabelPairs(lbsI, lbsJ); cmp != 0 {
+		return cmp < 0
+	}
+
+	// then compare timenanos, if they don't match return
+	if pI.TimeNanos < pJ.TimeNanos {
+		return true
+	} else if pI.TimeNanos > pJ.TimeNanos {
+		return false
+	}
+
+	// finally use ID as tie breaker
+	return bytes.Compare(pI.ID[:], pJ.ID[:]) < 0
+
+}
+
+func (s *profileStore) Flush(ctx context.Context) (numRows uint64, numRowGroups uint64, err error) {
+	if err := s.Close(); err != nil {
+		return 0, 0, err
+	}
+
+	if err := s.cutRowGroup(); err != nil {
+		return 0, 0, err
+	}
+
+	indexPath := filepath.Join(
+		s.path,
+		block.IndexFilename,
+	)
+
+	if err := s.index.writeTo(ctx, indexPath); err != nil {
+		return 0, 0, err
+	}
+
+	parquetPath := filepath.Join(
+		s.path,
+		s.persister.Name()+block.ParquetSuffix,
+	)
+
+	numRows, numRowGroups, err = s.writeRowGroups(parquetPath, s.RowGroups())
+	if err != nil {
+		return 0, 0, err
+	}
+
+	// cleanup row groups, which need cleaning
+	for _, rg := range s.rowGroups {
+		if err := rg.Close(); err != nil {
+			return 0, 0, err
+		}
+	}
+	s.rowGroups = s.rowGroups[:0]
+
+	return numRows, numRowGroups, nil
+}
+
+func (s *profileStore) prepareFile(path string) (closer io.Closer, err error) {
+	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o644)
+	if err != nil {
+		return nil, err
+	}
+	s.writer.Reset(file)
+
+	return file, err
+}
+
+// cutRowGroups gets called, when a patrticular row group has been finished and it will flush it to disk. The caller of cutRowGroups should be holding the write lock.
+// TODO: write row groups asynchronously
+func (s *profileStore) cutRowGroup() (err error) {
+	// do nothing with empty buffer
+	bufferRowNums := len(s.slice)
+	if bufferRowNums == 0 {
+		return nil
+	}
+
+	path := filepath.Join(
+		s.path,
+		fmt.Sprintf("%s.%d%s", s.persister.Name(), s.rowsFlushed, block.ParquetSuffix),
+	)
+
+	fileCloser, err := s.prepareFile(path)
+	if err != nil {
+		return err
+	}
+	defer runutil.CloseWithErrCapture(&err, fileCloser, "closing row group segment file")
+
+	// order profiles properly
+	sort.Slice(s.slice, s.profileSort)
+
+	n, err := s.writer.Write(s.slice)
```

Comment:
Great !

This opens the opportunity to improve serialization later (by writing []parquet.Row)